scipy/sparse/dok.py

```diff
@@ -351,122 +359,104 @@ def __radd__(self, other):
+                                      ' supported.')
         new = dok_matrix(self.shape, dtype=self.dtype)
-        for key in self.keys():
-            new[key] = -self[key]
-
+        dict.update(new, ((k, -self[k]) for k in iterkeys(self)))
         return new
 
     def _mul_scalar(self, other):
         res_dtype = upcast_scalar(self.dtype, other)
         # Multiply this scalar by every element.
         new = dok_matrix(self.shape, dtype=res_dtype)
-        for (key, val) in iteritems(self):
-            new[key] = val * other
+        dict.update(new, ((k, v * other) for k, v in iteritems(self)))
         return new
 
     def _mul_vector(self, other):
         # matrix * vector
-        result = np.zeros(self.shape[0], dtype=upcast(self.dtype,other.dtype))
-        for (i,j),v in iteritems(self):
+        result = np.zeros(self.shape[0], dtype=upcast(self.dtype, other.dtype))
+        for (i, j), v in iteritems(self):
             result[i] += v * other[j]
         return result
 
     def _mul_multivector(self, other):
         # matrix * multivector
-        M,N = self.shape
-        n_vecs = other.shape[1]  # number of column vectors
-        result = np.zeros((M,n_vecs), dtype=upcast(self.dtype,other.dtype))
-        for (i,j),v in iteritems(self):
+        M, N = self.shape
+        n_vecs = other.shape[1]
+        result = np.zeros((M, n_vecs), dtype=upcast(self.dtype, other.dtype))
+        for (i, j), v in iteritems(self):
             result[i,:] += v * other[j,:]
         return result
 
     def __imul__(self, other):
         if isscalarlike(other):
-            # Multiply this scalar by every element.
-            for (key, val) in iteritems(self):
-                self[key] = val * other
-            # new.dtype.char = self.dtype.char
+            dict.update(self, ((k, v * other) for k, v in iteritems(self)))
             return self
-        else:
-            return NotImplemented
+        return NotImplemented
 
     def __truediv__(self, other):
         if isscalarlike(other):
             res_dtype = upcast_scalar(self.dtype, other)
             new = dok_matrix(self.shape, dtype=res_dtype)
-            # Multiply this scalar by every element.
-            for (key, val) in iteritems(self):
-                new[key] = val / other
-            # new.dtype.char = self.dtype.char
+            dict.update(new, ((k, v / other) for k, v in iteritems(self)))
             return new
-        else:
-            return self.tocsr() / other
+        return self.tocsr() / other
 
     def __itruediv__(self, other):
         if isscalarlike(other):
-            # Multiply this scalar by every element.
-            for (key, val) in iteritems(self):
-                self[key] = val / other
+            dict.update(self, ((k, v / other) for k, v in iteritems(self)))
             return self
-        else:
-            return NotImplemented
+        return NotImplemented
+
+    def __reduce__(self):
```

Comment:
I think one can get it to work with ``__get(set)state__`` but also ``__new__`` and ``__getnewargs__`` would need to be implemented. Reason is that ``__setstate__`` is not called because ``__setitem__`` is called before it and it does not have ``shape`` variable hence everything breaks. If you know some other way to achieve this let me know, but I think this one is reasonably robust since it just redirects work to ``dict``.